File: Java/Tree/BJAlgo_5639.java
import java.io.*;
import java.util.*;

public class BJAlgo_5639 {
    static List post = new ArrayList();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String s = "";
        while ((s = br.readLine()) != null && (s.length() != 0)) {
            int num = Integer.parseInt(s);
            post.add(num);
        }

        if (post.size() > 0) {
            postOrder(0, post.size() - 1);
        }
    }

    static void postOrder(int start, int end) {
        if (start > end) {
            return;
        }

        int division = end + 1;
        for (int i = start + 1; i < end + 1; i++) {
            if ((int) post.get(start) < (int) post.get(i)) {
                division = i;
                break;
            }
        }

        postOrder(start + 1, division - 1);
        postOrder(division, end);
        System.out.println(post.get(start));
    }
}

File: Python/Tree/BJAlgo_5639.py
from sys import stdin
import sys
sys.setrecursionlimit(10**9)  # 재귀 허용 깊이 지정 (런타임 에러 방지)


def postorder(start, end):
    if start > end:
        return

    division = end + 1  # 나눌 위치 (root 보다 큰 처음 값 위치)
    for i in range(start + 1, end + 1):
        if post[start] < post[i]:
            division = i
            break

    postorder(start + 1, division - 1)  # 분할 왼쪽 (root 보다 작은 그룹)
    postorder(division, end)            # 분할 오른쪽 (root 보다 큰 그룹)
    print(post[start])


post = []
count = 0
while count <= 10000:
    try:
        num = int(stdin.readline())
    except:
        break

    post.append(num)
    count += 1

postorder(0, len(post) - 1)